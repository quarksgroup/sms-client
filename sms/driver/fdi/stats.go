package fdi

import (
	"context"
	"fmt"
	"time"

	"github.com/quarksgroup/sms-client/sms"
)

type statsService struct {
	client *wrapper
}

func (s *statsService) At(ctx context.Context, at string) (*sms.Stats, *sms.Response, error) {
	endpoint := fmt.Sprintf("stats/%s", at)
	out := new(stats)
	res, err := s.client.do(ctx, "GET", endpoint, nil, out)
	return convertStats(out), res, err
}

func (s *statsService) Current(ctx context.Context) (*sms.Stats, *sms.Response, error) {
	endpoint := "stats/now"
	out := new(stats)
	res, err := s.client.do(ctx, "GET", endpoint, nil, out)
	return convertStats(out), res, err
}

type stats struct {
	Success bool      `json:"success"`
	Date    time.Time `json:"date"`
	MT      outBox    `json:"mt"`
	MO      inBox     `json:"mo"`
}

type outBox struct {
	Sent      int64 `json:"sent"`
	Delivered int64 `json:"delivered"`
	Failed    int64 `json:"failed"`
	Pending   int64 `json:"pending"`
}

type inBox struct {
	Recieved int64 `json:"received"`
}

func convertStats(from *stats) *sms.Stats {
	return &sms.Stats{
		Sent:      from.MT.Sent,
		Delivered: from.MT.Delivered,
		Pending:   from.MT.Pending,
		Failed:    from.MT.Failed,
		Received:  from.MO.Recieved,
		Date:      from.Date,
	}
}
