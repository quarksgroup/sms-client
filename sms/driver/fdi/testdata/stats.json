{
    "success": true,
    "date": "2019-01-01T00:00:00.000Z",
    "mt": {
      "sent": 400,
      "delivered": 100,
      "failed": 50,
      "pending": 250
    },
    "mo": {
      "received": 420
    }
  }