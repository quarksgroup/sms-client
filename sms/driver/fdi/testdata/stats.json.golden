{
    "Date": "2019-01-01T00:00:00.000Z",
    "Sent": 400,
    "Delivered": 100,
    "Failed": 50,
    "Pending": 250,
    "Received": 420
}